### PR TITLE
Avatar entities

### PR DESCRIPTION
**Pull request by sethalves**
- Entities can be associated with an owning avatar (avatarEntities)
- avatarEntities live in local interface octrees but not in the entity-server octree
- avatarEntities are sent to other interfaces via the avatar-mixer rather than entity-server
- avatarEntities are saved to interface Settings and loaded from there on startup

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0ab0409979b20c39ca61dbf6d1a46c2d87e66df0 available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

File: libraries/entities/src/EntityEditPacketSender.cpp [hunk 1]
```diff
@@ -35,18 +36,54 @@ void EntityEditPacketSender::adjustEditPacketForClockSkew(PacketType type, QByte
...
-void EntityEditPacketSender::queueEditEntityMessage(PacketType type, EntityItemID modelID,
-                                                                const EntityItemProperties& properties) {
+void EntityEditPacketSender::queueEditEntityMessage(PacketType type,
+                                                    EntityTreePointer entityTree,
+                                                    EntityItemID entityItemID,
+                                                    const EntityItemProperties& properties) {
     if (!_shouldSend) {
         return; // bail early
     }
+    if (properties.getClientOnly()) {
```
Comment (ZappoMan) [inline libraries/entities/src/EntityEditPacketSender.cpp]: This feels like the wrong place to put this. The `EntityEditPacketSender`'s only job should be to send edit packets to the entity server... at least that's how I originally designed it. Maybe this should be in the `EntitySciptingInterface`? That would make more sense to me.

File: interface/src/avatar/Avatar.cpp [hunk 2]
```diff
@@ -160,6 +161,89 @@ void Avatar::animateScaleChanges(float deltaTime) {
...
+            }
+        }
+    });
+
+    AvatarEntityIDs recentlyDettachedAvatarEntities = getAndClearRecentlyDetachedIDs();
+    foreach (auto entityID, recentlyDettachedAvatarEntities) {
+        if (!_avatarEntityData.contains(entityID)) {
+            EntityItemPointer dettachedEntity = entityTree->findEntityByEntityItemID(EntityItemID(entityID));
+            if (dettachedEntity) {
+                // this will cause this interface to listen to data from the entity-server about this entity.
```
Comment (ZappoMan) [inline interface/src/avatar/Avatar.cpp]: Just to clarify the "spec" here. If I detatch an entity in a domain, then I leave it behind? So it becomes a static entity in the domain? That's a cool feature... I like it. Where do we call the "add"?

Comment (sethalves) [inline interface/src/avatar/Avatar.cpp]: That's how the previous iteration of this worked, but I kept running into trouble.  This time I wont allow entities to move back and forth -- to make the transition, a script will have to delete one type and create the other.  I'll fix the comment.

Comment (ZappoMan) [inline interface/src/avatar/Avatar.cpp]: But where are you deleting it? It looks like from this code, the entity is going to end up being abandoned in the local tree... which would be not so good... it would still render on machines that previously saw it, etc... but not be visible to other/new users...

**Comment by hifi-gustavo** [comment]
Build console output for commit: c572e1dc3a837e0c1f0d42678ba0ba67591697b6 available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: f90b9ec65dc58f06dc9d94e7e915a077231d0526 available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1e849956c9ed74dd9156e44f94173b2a4bda9263 available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4b13fd969e0609e9abc5ac8e08daf92835f19e7e available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b05ab1b17e219186e2c8c64d122b64e41b9ca342 available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 144715f00cf7b8bb33ed551d9f57e4fed5859051 available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 356ccdba26cb668e999c4af45dc691f134b17f8f available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by sethalves** [comment]
to test:
- smoke test
- experiment with passing addEntity a clientOnly flag of true, for example: open the console and run

  Entities.addEntity({type: 'Box', position: MyAvatar.position}, true);

- log out and back in, see that the box come back.  Other interfaces (must be from this PR) will see it also
- delete it, log out, log back in, make sure it doesn't return

**Comment by hifi-gustavo** [comment]
Build console output for commit: 872f1b0c64d9dc7479dbc1a9de3615263dc21818 available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9038647b6536955813f5757316a15bba317b5211 available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: ac506dc4b77de0bd8831796de3ffe6fad855a6c9 available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 032fbd9a55b6ff220c7ac0e17f989ded353bae6d available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2101d7dea3e9b71029aa585571baa36cb2dd0182 available at: https://gist.github.com/4054d8337e12e5a1ddd5b2f640b8aaa4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2f93a38a6a52e294ff89b92485fed18dd9ef7880 available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: eb19c5be13d6b2c5c147e18ccdba8e1d4e0191de available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: dabecb7cc29977e93fcc466a9738da57b39bde40 available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by sethalves** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: dabecb7cc29977e93fcc466a9738da57b39bde40 available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by sethalves** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: dabecb7cc29977e93fcc466a9738da57b39bde40 available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 81af0d0a47b8fdab71bb933bde630b05c422b72c available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0ab987faac6a56a0f36adab0bae74ca2dc4bcad6 available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: efe02c0fa120666b5296d0e559e272b67b2fb412 available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: bfac7c6f68f008d29cf83d8275585b60bf352b86 available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 880b52aad288ec91d73fdfc478acbe57a110eb3e available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by OmegaHeron** [comment]
Tested Windows 10 64 bit - i5 6500 CPU - GTX970 GPU - **Desktop mode - keyboard/mouse**
All standard tests pass.  No negative delta seen for CPU/GPU/RAM/FPS stats.

Running this PR to current RC (or latest DEV) Stack - using console command - creates cube that persists - using production Interface you can see cube after relog.  So - this PR with that command on current servers creates entity in the "classic" way it seems.

Running this PR to this PR Stack (on Linux cloud droplet) - use console command to create entity - it creates and is present on relog.  Using production Interface you do not set entity validating it's in "different" space.

Using current Interface to this RC stack and running console command creates entity that persists.  This one can be seen by both PR and RC Interfaces so it's existing in "classic" space.  Running this PR and logging in I can see entities created in either octree.

Second AV seems to see whatever first AV creates within capability of its running interface level.

I'd be more comfortable if a few others tested this as it can be so hit/miss on how others machines/configs react to these things.  But.  QA Pass for me.

**Comment by OmegaHeron** [comment]
Hold on QA Pass please - I need to look at one additional thing and determine if it's bug or my mistake.

**Comment by OmegaHeron** [comment]
tldr version
Use this pr Interface and stack.  Create a cube using console command.  Copy path so you have coordinates.  Change to other domains and navigate to path you copied.  See cube?  I do.

Long version
Ok - running this PR I went to a domain running current RC Stack - I used console command to create an entity - It appeared.  I changed to another domain and same cube exists in it as well.  I moved to playa then put my path to location where I could see cube in my own domains and see it there as well.

Not sure if that's bug or simply Interface being unable to deal with its connecting to servers not providing the alternate octree.

Created another entity in domain running this PR - made it a distinct color and I see it at same location in any other domain I move to as well.

Again - I'm unsure if this is really a bug or not as it could be vestige of not having full compat between this PR and old Stacks.

**Comment by OmegaHeron** [comment]
QA Pass for me - after speaking with @sethalves this is behaving exactly as intended.

**Comment by ZappoMan** [comment]
@OmegaHeron - the client in this PR can not talk to servers from a different PR... so it's expected that the feature wouldn't work when you connected to playa.

**Comment by OmegaHeron** [comment]
Well - @ZappoMan - it can, it just doesn't do it well.  :)  It's not like the old days where it would refuse to connect.  This PR happily connects to current RC and DEVEL level stacks.

**Comment by ZappoMan** [comment]
That's related to a separate (arguably a bug) that changes to the avatar
mixer protocol don't seem to complain to the user. The entity protocol
didn't change... so you do see entities. This is a feature of the avatar
mixer only.

File: interface/src/avatar/Avatar.cpp [hunk 3]
```diff
@@ -157,6 +170,98 @@ void Avatar::animateScaleChanges(float deltaTime) {
...
+            QVariantMap asMap = variantProperties.toMap();
+            QScriptValue scriptProperties = variantMapToScriptValue(asMap, scriptEngine);
+            EntityItemProperties properties;
+            EntityItemPropertiesFromScriptValueHonorReadOnly(scriptProperties, properties);
+            properties.setClientOnly(true);
+            properties.setOwningAvatarID(getID());
+
+            // there's no entity-server to tell us we're the simulation owner, so always set the
+            // simulationOwner to the owningAvatarID and a high priority.
+            properties.setSimulationOwner(getID(), 129);
```
Comment (AndrewMeadows) [inline interface/src/avatar/Avatar.cpp]: Simulation ownership priority 129 should be given a name and declared as a constant in `SimulationOwner.h`.

**Comment by hifi-gustavo** [comment]
Build console output for commit: cf829f1babf58f4d442bdebdb8696f2c7f164fdf available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 419c744bd4f0f54910cc22781771a4721d9f8a0e available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by sethalves** [comment]
I hate you gustavo

**Comment by sethalves** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 419c744bd4f0f54910cc22781771a4721d9f8a0e available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by sethalves** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 419c744bd4f0f54910cc22781771a4721d9f8a0e available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by sethalves** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 419c744bd4f0f54910cc22781771a4721d9f8a0e available at: https://gist.github.com/4054d8337e12e5a1ddd5b2f640b8aaa4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1de9598243c4557d80477037fb5af863b04f1864 available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by sethalves** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1de9598243c4557d80477037fb5af863b04f1864 available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by sethalves** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1de9598243c4557d80477037fb5af863b04f1864 available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by sethalves** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1de9598243c4557d80477037fb5af863b04f1864 available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

File: interface/src/avatar/Avatar.cpp [hunk 4]
```diff
@@ -157,6 +170,98 @@ void Avatar::animateScaleChanges(float deltaTime) {
...
+    EntityTreePointer entityTree = treeRenderer ? treeRenderer->getTree() : nullptr;
+    if (!entityTree) {
+        return;
+    }
+
+    bool success = true;
+    QScriptEngine scriptEngine;
+    entityTree->withWriteLock([&] {
+        AvatarEntityMap avatarEntities = getAvatarEntityData();
+        qDebug() << "---------------";
```
Comment (ZappoMan) [inline interface/src/avatar/Avatar.cpp]: should we remove this debugging?? is it log spam?

File: interface/src/avatar/Avatar.cpp [hunk 5]
```diff
@@ -157,6 +170,98 @@ void Avatar::animateScaleChanges(float deltaTime) {
...
+            properties.setSimulationOwner(getID(), AVATAR_ENTITY_SIMULATION_PRIORITY);
+
+            if (properties.getParentID() == AVATAR_SELF_ID) {
+                properties.setParentID(getID());
+            }
+
+            EntityItemPointer entity = entityTree->findEntityByEntityItemID(EntityItemID(entityID));
+
+            if (entity) {
+                qDebug() << "avatar-entities existing entity, element =" << entity->getElement().get();
```
Comment (ZappoMan) [inline interface/src/avatar/Avatar.cpp]: is this log spam?

**Comment by ZappoMan** [comment]
Dev Approved pending Seth fixing the debug log spam.

**Comment by hifi-gustavo** [comment]
Build console output for commit: d95d3ff3ac61c722a109fe50fe99859b31eee585 available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by ZappoMan** [comment]
That's gonna leave a mark!

**Comment by hifi-gustavo** [comment]
Build console output for commit: f91a2403be6a75a7aa13ffc3e23474385d7edef8 available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 637735bbc3e6099ed6648d31812a027d866ffcfb available at: https://gist.github.com/02a6b2933cf2f57c1da00946d813af9c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7840.dmg
